bot/utils/services.py

```diff
@@ -57,5 +112,142 @@ async def send_to_paste_service(contents: str, *, extension: str = "") -> Option
+            return _UnfurlReturn(data["destination"], data["depth"], None, expiry - CACHE_LENGTH)
+
+
+async def _unfurl_url(url: str, redirects: int, continues: int, max_continues: int) -> _JOIN_RETURNS:
+    """
+    The actual core logic of the unfurling.
+
+    Unlike the public utility, this will only follow one redirect, and parse the results.
+    The handling of the results is managed by the public function.
+
+    Returns a tuple with a final UnfurlReturn and response status,
+    or a tuple containing data to allow the caller to continue, or None.
+    """
+    # See link for documentation on how to use the worker
+    # https://github.com/python-discord/workers/tree/main/url-unfurler
+    response = await _attempt_request(URLs.unfurl_worker, 3, json={"url": url}, raise_for_status=False)
+
+    if response is None:
+        return
+
+    if response.status == 200:
+        # Success, return the destination
+        content = await response.json()
+        now = datetime.datetime.now(tz=datetime.timezone.utc)
+        return _UnfurlReturn(content["destination"], redirects + content["depth"], None, now), response.status
+
+    elif response.status == 400:
+        # Unrecoverable error with the URL used for this unfurling
+        error_message = (await response.json())["error"]
+        log.warning(f"Failed to unfurl URL ({url}) with error message: {error_message}")
+        return _UnfurlReturn(error=error_message), response.status
+
+    elif response.status == 416:
+        # Max depth reached by the worker, try again or exit
+        data = await response.json()
+
+        if continues < max_continues:
+            # Try again (the caller will make a new call with the next URL)
+            log.debug(
+                f"Hit maximum depth while unfurling URL ({url}) after {data['depth']} attempts. "
+                f"Last request was made to ({data['final']}), and next request will be made to {data['next']}. "
+                f"{continues + 1}/{max_continues} failures."
+            )
+
+            return data["next"], data["depth"]
+
+        else:
+            # Give up
+            redirects += data["depth"]
+            log.info(f"Failed to unfurl URL ({url}) after {redirects} redirects.")
+            return _UnfurlReturn(data["next"], redirects, data["error"]), response.status
+
+    elif response.status == 418:
+        # The redirect chain has a broken link, can not continue
+        data = await response.json()
+        redirects += data["depth"]
+
+        log.warning(
+            f"Failed to unfurl URL ({url}) due to a broken link in the chain. "
+            f"Followed {redirects} redirects, and stopped at ({data['final']})."
+        )
+        return _UnfurlReturn(data["final"], redirects, data["error"]), response.status
+
+    else:
+        response.raise_for_status()
+
+
+async def unfurl_url(url: str, *, max_continues: int = 0, use_cache: bool = True) -> Optional[_UnfurlReturn]:
```

Comment:
I don't understand `max_continues`. I passed the value 1 in the command with the url `https://pydis.com/coc` but it still resolved the *2* redirects?

Reply:
`max_continues` is used when we hit the maximum redirection depth, which is hardcoded to 3 in L151:
https://github.com/python-discord/bot/blob/80abaef2cb991a86e97a40650efdc7b9541341d2/bot/utils/services.py#L151

That is to say, if you have a site that resolves after 5 redirects, the program will work as follows:
- `0 continues`: Will not resolve since we only do a max of 3 redirects
- `1 continue`: Will fail once, the continue. The max redirects for this is 6, and since we only have 5, it'll resolve.

I see why this could be confusing if you don't have knowledge of the worker though, any ideas on what we could do?

Reply:
Maybe call it `attempt`s? Since what's actually happening is that we're attempting to resolve the URL that many times. To me (and I would really assume to many others) `max_continues` maps to `redirects`. `attempt` shares the same downsides though I suppose..

Why can't `max_continues` be passed in there directly (if we give it sensible limits)?

Reply:
This is a little better documented on the actual command:
> Setting `max_continues` will continue unfurling, even if we hit limits on the worker.

I don't mind renaming it if we find a better name. The reason I didn't bother making the other number configurable is that the range of valid values for it is a lot more difficult to control, and is dictated entirely by the range in the worker. I didn't want to maintain the range in two places. Continues is completely independent of the worker though, which removes that problem.

Reply:
I guess it isn't too bad. Could just manually define the `continues` based on the redirects % max, and just have the max defined in config. That way the command can just have the raw number like you were trying initially.

Reply:
> This is a little better documented on the actual command:
> > Setting `max_continues` will continue unfurling, even if we hit limits on the worker.

Yeah that isn't clear to someone who doesn't understand the workers. This would be better:
> Setting `max_continues` means that unfurling will retry if the limit of redirects on an individual worker is hit.

I think what makes sense is just having the `continues` yes..

Reply:
32eea072